expose parameters via config-map

stefanhengl opened the pull request:
So far the subscription-id was the only accepted parameter.  With this PR we can configure the stream by supplying a configuration map to `consume-subscription`.

Commit:
expose parameters via config-map

dryewo commented on project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject me.dryewo/clj-nakadi-java "0.0.2"
+(defproject me.dryewo/clj-nakadi-java "0.0.3"
```
Please keep the version number the same, it will be bumped as part of the release procedure.

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,18 +47,36 @@
+      ))
+
+
+(defn consume-subscription [client config-map callback]
```
The previous version was accepting either a subscription-id or a full `StreamConfiguration` object.
I would rather keep it that way, but expose `->subscription-stream-config` as a helper function to convert a config map to such an object, so that the user would call it like this:

```clj
(consume-subscription client (->subscription-stream-config {...}) callback)
```

An alternative option would be to check if the argument is a map and then convert to this kind of object implicitly, making it 3 different possibilities instead of 2:
* string (sibscription ID)
* config map (will be converted to `StreamConfiguration` automatically)
* instance of `StreamConfiguration`

stefanhengl replied:
I feel it is a cleaner approach to just accept  a `StreamConfiguration` object. `->subscription-stream-config` now serves as a helper function just as you suggested.

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,18 +47,36 @@
+
+(defn ->subscription-stream-config [config-map]
+  (-> (StreamConfiguration.)
+      (set-param config-map :subscription-id #(.subscriptionId % %2) true)  ;; required
```
I would rather programmatically convert `:subscription-id` to `subscriptionId` method calls to not hardcode specific method here.
Something like this:

```clj
(let [stream-config (StreamConfiguration.)]
  (doseq [[k v] config-map]
    (let [method (convert-keyword-to-method k)] ; should throw an assertion error when method does not exist
      (call-method-via-reflection stream-config method v))
  stream-config)
```

stefanhengl replied:
Done. However, I decided not to throw an exception if the method does not exist. I feel it is more  resillient this way.

dryewo commented on README.md:
```diff
@@ -46,7 +47,7 @@ Recommended way:
 (m/defstate consumer
   :start (do
-           (nakadi/consume-subscription @client subscription-id callback)
+           (nakadi/consume-subscription @client {:subscription-id subscription-id :batch-limit batch-limit} callback)
```
I would provide the examples of all 3 options here: string, map and `StreamConfiguration`

Commit:
fixes issue from CR

- consume-subscription accepts stream-config only
- method names are determined programmatically
- update Readme
- version number reset to 0.0.2

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,37 +48,35 @@
-   (set-param StreamConfiguration config-map key method false)))
+
+(defn ->camel-case [kebab]
+  (let [a (clojure.string/split (name kebab) #"\-")]
```
Please use self-descriptive variable name instead of just `a`. `name-parts`, perhaps?

stefanhengl replied:
I followed your suggestion.

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,37 +48,35 @@
+ 
+
+(defn str-invoke [instance method-str & args]
+  (clojure.lang.Reflector/invokeInstanceMethod
```
IMHO this would also look ok if you put all things on the same line

stefanhengl replied:
I agree.

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,19 +48,35 @@
+(defn set-param [streamConfigInstance k v]
+  (let [method (->camel-case k)]
+    (if (method-exists? method)
+      (str-invoke streamConfigInstance method v)
```
does str-invoke always return streamConfigInstance? if it's not, then the following calls will break

dryewo replied:
I'd rather

```clj
(defn set-param [streamConfigInstance k v]
  (let [method (->camel-case k)]
    (when (method-exists? method)
      (str-invoke streamConfigInstance method v))
    streamConfigInstance))
```

stefanhengl replied:
You are right, this way is safer.

dryewo commented on src/clj_nakadi_java/core.clj:
```diff
@@ -47,19 +48,35 @@
+  (reduce-kv set-param (StreamConfiguration.) config-map))
+
+
+(defn consume-subscription [client stream-config callback]
```
Removing the support for the string input would break existing use cases, I'd rather keep it.
Also, raising an assertion error if subscription ID is missing should be done here.
IMHO failing early is a good thing.

stefanhengl replied:
If we have to support multiple input formats, then I'd rather go with all three: map, string, or an instance of `StreamConfiguration`. I changed the code and implemented `consume-subscription` as multimethod.

dryewo reviewed (changes requested):
Could you please cover `->subscription-stream-config` with unit tests? This shouldn't be hard, but will raise the confidence significantly.

Commit:
allow various input formats

- config can either be a string, a map, or an instance of StreamConfiguration.
- Assertion errors are thrown if :subscription-id is not part of config-map or if the configuration is supplied in an unsupported format
- minor formatting changes

dryewo commented:
Looks good to me. Thanks for the contribution! I'll release the next version shortly.
